WIP climber subsystem extenders

BeastLaptop07 opened the pull request:
Need to verify encoder values and speed the climber should move at.
Need to figure out CANSparkMax can command position, currently using set soft position limit and set(double speed)
PR open by: Nithila Balaji, Emily Enright, Maddy Paradis

tl edit: 2/5 added todo's from the code.
- [ ] Split into separate commajds
- [ ] State diagram
- [ ] Fully integrate (i.e. extend then pull back)

dlaufen committed:
Suggestions and TODOs

jstanlick committed:
Mechanum First steps

Build in progress

jstanlick committed:
Forgot 3 motor definitions

dlaufen committed:
Added MechanumDriveCmd, fixed compile bugs

jstanlick committed:
Replaced SwerveDriveTrain with Mecanum

Not tested

kamith committed:
Working Mecanum Drive

Changed the Motor Orientation in Constants.java to match the Physical Robot

sarah-i-rubenstein committed:
More variables

dlaufen committed:
Naming convention, SparkMax PID init

sarah-i-rubenstein committed:
Auto stash before merge of "Climber" and "origin/Climber"

dlaufen committed:
Restored PID initialization, added notes/todo

BeastLaptop07 committed:
climber subsystem extenders

BeastLaptop07 committed:
climber networktable

tomlazar commented on src/main/java/frc/robot/commands/Climbernetwork.java:
```diff
@@ -0,0 +1,38 @@
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+public class Climbernetwork extends CommandBase {
```
```suggestion
public class TestClimber extends CommandBase {
```

This should eventually be able to run the climber in the simpliest way possible, if it becomes to complex to run just over network tables, then a series of commands should be used, or even switching to "production" quality commands.

tomlazar commented on src/main/java/frc/robot/commands/Climbernetwork.java:
```diff
@@ -0,0 +1,38 @@
+package frc.robot.commands;
```
```suggestion
package frc.robot.commands.test;
```

should be in the test folder.

tomlazar commented on src/main/java/frc/robot/commands/Climbernetwork.java:
```diff
@@ -0,0 +1,38 @@
+    private Double last = 0.0;
+
+    // constructor to create network table
+    Climbernetwork(){
```
match class name

aaron97neu commented on src/main/java/frc/robot/commands/Climbernetwork.java:
```diff
@@ -0,0 +1,38 @@
+
+    }
+    @Override
+    public void execute(){
```
Needs to be finished. Look at this file for a rough example: https://github.com/2202Programming/FRC2022/blob/fullshooter/src/main/java/frc/robot/commands/test/TestMotorCommand.java

tomlazar commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -0,0 +1,118 @@
+    public void setExtension(double inches) {
+        float count =  (float) inches; // convert command inches to encoder counts
+        SoftLimitDirection direction = null ;
+        double speed = 0 ;
```
Rather then leaving this in a open loop control, we prefer to use PID loops and closed control systems running on the actual device for a more fine grained control.

For a example of a position PID loop see this [demo project](https://github.com/REVrobotics/SPARK-MAX-Examples/tree/master/Java/Position%20Closed%20Loop%20Control) by REV. Specifically this [line](https://github.com/REVrobotics/SPARK-MAX-Examples/blob/c18eb6f15f9d9b8d371da83e7302470f0d50c170/Java/Position%20Closed%20Loop%20Control/src/main/java/frc/robot/Robot.java#L112).

aaron97neu replied:
@dlaufen Could you link some example code that shows how to change from native units (rotations/pulses?) to sane (inches in this case). Would be a nice complement to add while making this change

tomlazar reviewed (changes requested):
@dlaufen we should probably remove the Mecanum parts at some point.

aaron97neu commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -0,0 +1,118 @@
+    public boolean readyToTraverse() {
+        return false;
+    }
+    // @param inches from extender current position
```
We need to reconsider if it is smarter to base our 0 off the current position or fully retracted. I am starting to think 0 being set at fully retracted is the best bet. All values should be positive in this case, need to enforce that.

aaron97neu reviewed (changes requested):
Did review with Tom, agree with his requested changes. Added a few of my own

nathanaelren committed:
WIP climber changes. Moved NT to subsystem, created command

nathanaelren committed:
WIP. Climber changed to do all climbs at once, added conditions

Still need to check for rotation

nathanaelren committed:
WIP fixed some logic errors

nathanaelren committed:
Added todos

aaron97neu commented:
State diagram brainstormed 2/7
![PXL_20220208_010444762.jpg](https://user-images.githubusercontent.com/13039768/152898469-1200dbf0-4d5b-4e56-938f-5853c740e25f.jpg)

nathanaelren committed:
WIP mid stuff first crack

Commit:
made climber subsytem extend subsystem base, rewrote climb command to function as a sequential command wrapper

nathanaelren committed:
WIP hopefully all commands necessary

nathanaelren committed:
Merge remote-tracking branch 'origin/Climber' into Climber

nathanaelren committed:
WIP removed some red

nathanaelren committed:
no more relative

nathanaelren committed:
Moved final vars to constants file

nathanaelren committed:
Commented out warnings to make everything look clean and make me happy

nathanaelren committed:
moved everything back to execute for pid etc

Commit:
Roation with relays, need to switch to pwm + single channel encoders

aaron97neu committed:
Merge branch 'Climber' of github.com:2202Programming/FRC2022 into Climber

tomlazar commented:
@aaron97neu 

```java

class ArmRotation {
    private Counter m_counter;
    private PWMMotorController m_motor;
    private double speed;
    private double tolerence;

    private int absPositon = 0;
    private int desPositon = 0;
    private Boolean kForward = true;

    public ArmRotation(Counter m_counter, PWMMotorController m_motor, double speed, double tolerence) {
        this.m_counter = m_counter;
        this.m_motor = m_motor;
        this.speed = speed;
        this.tolerence = tolerence;
    }


    public void periodic() {
        // add the adjustment
        double factor = 1;
        if (!kForward) factor = -1;
        absPositon += factor * m_counter.get();
        m_counter.reset();

        if(Math.abs(desPositon - absPositon ) > tolerence) {
            m_motor.set(factor * speed);
        } else {
            m_motor.set(0);
        }
    }

    public void set(int desired) {
        kForward = desired > absPositon;
        desPositon = desired;
    }
}
```

beastsoftware2202 committed:
abstracted the arm rotation

 switched from the relay to the spark motor controller
-Maddy paradis, Emily Enright

tomlazar committed:
ffeat: move table entries

aaron97neu commented:
Current status: 
- Climber Subsystem seems to be fully implemented. Still needs a second pass to clean stuff up. Specifically Extension routine should be abstracted to a class like ClimberRotation.
- Climber test file is created but needs to be implemented. Planning on finishing implementing and testing Wednesday with @nathanaelren if he is available

nathanaelren committed:
test command bound to driver button x

nathanaelren committed:
Test command switched to NTs

Co-authored-by: Aaron Neustedter <aaron@neustedter.net>

BeastLaptop07 committed:
Hotfixes to get proto working

nathanaelren committed:
updates

Co-authored-by: Aaron Neustedter <aaron@neustedter.net>

nathanaelren committed:
updates v2

Co-authored-by: Aaron Neustedter <aaron@neustedter.net>

nathanaelren committed:
logic changes

Co-authored-by: Aaron Neustedter <aaron@neustedter.net>

nathanaelren committed:
more logic changes

Co-authored-by: Aaron Neustedter <aaron@neustedter.net>

BeastLaptop07 committed:
changed rot speed of rot motors

nathanaelren committed:
logic stuff

Co-authored-by: Aaron Neustedter <aaron@neustedter.net>

beastsoftware2202 committed:
Work in progress

BeastLaptop07 committed:
wip

nathanaelren committed:
better testing v1

maddyp7 committed:
conversions for counts and degrees

maddyp7 committed:
Merge remote-tracking branch 'origin/Climber' into Climber

nathanaelren committed:
update for new nt entry

nathanaelren committed:
semantics

25enrige committed:
Switching back to cansparkmaxes for rotator, clean up constants

aaron97neu committed:
Fix merge conflict with setRot()

aaron97neu commented:
Suggested plan of attack for the new limit switches:

- Communicate with mechanical team and identify limit switches, their type, location, how they are wired and document. I think they are all magnetic but it should not matter if that changes

- Assuming they are wired to the canspark max's, use this example to implement them: https://github.com/REVrobotics/SPARK-MAX-Examples/blob/master/Java/Limit%20Switch/src/main/java/frc/robot/Robot.java. The comments explain how things work nicely. Ensure you are using `SparkMaxLimitSwitch` objects/types. Additional docs: https://codedocs.revrobotics.com/java/com/revrobotics/cansparkmax

- If they are wired back to the rio example code from last year can be found here: https://github.com/2202Programming/FRC-2019/blob/master/FRC-Robot-2019/src/main/java/frc/robot/subsystems/ClimberSubsystem.java

Once that is finished, Climber subsystem needs to have a good cleanup and everything needs to be tested

dlaufen commented:
Makes sense.  I expect limit switches will be directly wired to sparkmax.
On 2/22/2022 7:33 PM, Aaron Neustedter wrote:
>
> Suggested plan of attack for the new limit switches:
>
>  *
>
>     Communicate with mechanical team and identify limit switches,
>     their type, location, how they are wired and document. I think
>     they are all magnetic but it should not matter if that changes
>
>  *
>
>     Assuming they are wired to the canspark max's, use this example to
>     implement them:
>     https://github.com/REVrobotics/SPARK-MAX-Examples/blob/master/Java/Limit
>     Switch/src/main/java/frc/robot/Robot.java. The comments explain
>     how things work nicely. Ensure you are using |SparkMaxLimitSwitch|
>     objects/types. Additional docs:
>     https://codedocs.revrobotics.com/java/com/revrobotics/cansparkmax#getForwardLimitSwitch(com.revrobotics.SparkMaxLimitSwitch.Type)
>
>  *
>
>     If they are wired back to the rio example code from last year can
>     be found here:
>     https://github.com/2202Programming/FRC-2019/blob/master/FRC-Robot-2019/src/main/java/frc/robot/subsystems/ClimberSubsystem.java
>
> Once that is finished, Climber subsystem needs to have a good cleanup 
> and everything needs to be tested
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/2202Programming/FRC2022/pull/5#issuecomment-1048374358>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AC3I2ZIFJRU7FAD4DOX67L3U4Q2PPANCNFSM5NH4OKAQ>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
> or Android 
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>
> You are receiving this because you were mentioned.Web Bug from 
> https://github.com/notifications/beacon/AC3I2ZOUSNWSCJDDVLG4MHDU4Q2PPA5CNFSM5NH4OKA2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOHZ6OYVQ.gifMessage 
> ID: ***@***.***>
>
> [ { ***@***.***": "http://schema.org", ***@***.***": "EmailMessage", 
> "potentialAction": { ***@***.***": "ViewAction", "target": 
> "https://github.com/2202Programming/FRC2022/pull/5#issuecomment-1048374358","url": 
> "https://github.com/2202Programming/FRC2022/pull/5#issuecomment-1048374358", 
> "name": "View Pull Request" }, "description": "View this Pull Request 
> on GitHub", "publisher": { ***@***.***": "Organization", "name": "GitHub", 
> "url": "https://github.com" } } ]

nathanaelren committed:
partially implemented limit switches

    Co-authored-by: Alan Sudhan <23sudhaa@elmbrookstudents.org>
    Co-authored-by: Edric Zhou <24zhoue@elmbrookstudents.org>

nathanaelren committed:
partially implemented limit switches

Co-authored-by: Alan Sudhan <23sudhaa@elmbrookstudents.org>
Co-authored-by: Edric Zhou <24zhoue@elmbrookstudents.org>

nathanaelren committed:
Merge remote-tracking branch 'origin/Climber' into Climber

beastsoftwaredev committed:
Climber Testing

nathanaelren committed:
Added rot checks for isFinished and amp limits on all relevant commands

nathanaelren committed:
added method to check for isFinished in subsystem, not implemented

nathanaelren committed:
added generic command with inputs for desired rotation, extension

Not implemented anywhere

beastsoftwaredev committed:
wip cleaned up smartdashboard

tomlazar commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -295,29 +246,158 @@ public void periodic() {
+    private NetworkTableEntry nte_des_pos_deg;
+    private NetworkTableEntry nte_des_pos_count;
+    private NetworkTableEntry nte_curr_current_amp;
+    private NetworkTableEntry nte_curr_current_limit_amp;
```
you_can_really_tell_aaron_only_programs_python

tomlazar commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -295,29 +246,158 @@ public void periodic() {
+        nte_des_pos_count.setDouble(0);
+        nte_curr_current_amp = network_table.getEntry("Current Amperage (amps)");
+        nte_curr_current_amp.setDouble(0);
+        nte_curr_current_limit_amp = network_table.getEntry("Amperage Limit (amps)");
```
maybe group inputs/outputs under

`$root/input/whatever`
`$root/output/whatever`

tomlazar commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -295,29 +246,158 @@ public void periodic() {
+        nte_curr_current_limit_amp = network_table.getEntry("Amperage Limit (amps)");
+        nte_curr_current_limit_amp.setDouble(0);
+
+        CONVERSION_FACTOR = motor_rot.getEncoder().getPositionConversionFactor();
```
can this change when its runnin?

tomlazar commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -295,29 +246,158 @@ public void periodic() {
+        nte_curr_pos_count.setDouble(motor_curr_deg / CONVERSION_FACTOR);
+        nte_curr_current_amp.setDouble(motor_rot.getOutputCurrent());
+// current limit n stuff
+        motor_rot.setSmartCurrentLimit((int) nte_curr_current_limit_amp.getDouble(5));
```
can trafic, test for changes

tomlazar commented on src/main/java/frc/robot/commands/MidClimbRetract.java:
```diff
@@ -22,11 +23,11 @@ public void execute() {
+        return ((Math.abs(climber.getLeftExtEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_LENGTH)
+               && (Math.abs(climber.getRightExtEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_LENGTH)
+               && (Math.abs(climber.getRightRotEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_ROTATION)
+               && (Math.abs(climber.getLeftRotEncoder().getPosition()) <= Constants.ClimbSettings.TOLERANCE_LENGTH)
```
use the new subsystem method?

tomlazar commented on src/main/java/frc/robot/subsystems/Climber.java:
```diff
@@ -295,29 +246,158 @@ public void periodic() {
+    }
+
+    public boolean checkIsFinished(double ext_pos, double rot_pos) {
+        return(Math.abs(this.getLeftExtEncoder().getPosition() - ext_pos) <= Constants.ClimbSettings.TOLERANCE_LENGTH 
```
take tolerence in method, and move checks to extension/rotation subclass

aaron97neu committed:
Add comments to guide finishing refactoring

maddyp7 committed:
Cleaned up climber subsystem

maddyp7 committed:
added calibrate to climber

BeastLaptop07 committed:
they move now blame aaron for the rest

beastsoftwaredev committed:
xbox controls

beastsoftwaredev committed:
wip xbox buttons

beastsoftwaredev committed:
changes for calibration + testing, wip

TODO: fix encoder conversion stuff

nathanaelren committed:
numbers n stuff

BeastLaptop07 committed:
doozy of changes -- calibrate command now, fixed conversions (I think)

-Calibrate is now a command. Any reference to calibrate has been removed from all other classes.
-Did some conversion stuff, now deg and inch should be default units; should be accurate.
-Tested mid climb command but speed could be tamed. Higher climb -- extension works well. but for some reason rotation won't stop. NTs show desired rotation exists, and yet motor blows right past it.
Co-authored by: Nathanael Ren <nren2013@gmail.com>
Co-authored by: Carly Janscha <23janscc@elmbrookstudents.org>

dlaufen committed:
Code-review cleanup.  Need to test units.

dlaufen committed:
set not get of a nte

dlaufen committed:
scaling, use naturalunits

dlaufen committed:
scaling rotation

dlaufen committed:
WIP pos/vel modes, compensation, minor name edits

beastsoftwaredev committed:
climber extension pid tuning

dlaufen committed:
masterful copy-pasta

dlaufen committed:
cleaned zeroing arms

dlaufen committed:
duh!!!!

dlaufen committed:
nte for velocity

dlaufen committed:
arbFF for rotation

beastsoftwaredev committed:
rot pid tuning

dlaufen committed:
wip

dlaufen committed:
pit tools

dlaufen committed:
WIP testing

beastsoftwaredev committed:
consts

beastsoftwaredev committed:
Merge remote-tracking branch 'origin/Climber' into Climber

dlaufen committed:
wip:pit test command

beastsoftwaredev committed:
pitalign works

dlaufen committed:
Simple climb cmd testing

dlaufen committed:
buttons

beastsoftwaredev committed:
wip wrapped some pids and did some testing

dlaufen committed:
WIP: outloop independent, removed unused code, cleanups

totally untested

dlaufen committed:
Options for MoveArms

dlaufen committed:
test arm sync

beastsoftwaredev committed:
climber

beastsoftwaredev committed:
build climb cmd

dlaufen committed:
wip

dlaufen committed:
nt names, comments on cmd

dlaufen committed:
wip

dlaufen committed:
current limit cleanups, rotation current and dutycyc to nte

dlaufen committed:
new CAN ID, current limit brushed, consts
